docs/DesignDocs/SGX_0-base_Enclave.md

```diff
@@ -0,0 +1,228 @@
+
+If the 0-page is within the SGX Enclave linear address range, and no Enclave EPC
+page is committed at address 0, the 0-page becomes an Enclave Guard Page. Even
+if a malicious OS creates a Page Table entry for the guard pages, Enclave code
+access to the guard pages will always trigger a Page Fault. If the untrusted SGX
+Enclave loader EADDs Enclave page to replace the Guard Pages, or does not set
+`SECS.BASEADDR` to 0, as expected, the Enclave measurement will be different
+from the expected value and the attack will be detected.
+
+To include the 0-page within the SGX Enclave linear address range, SECS.BASEADDR
+needs to be set as 0. Effectively, all pages with linear address lower than the
+minimum linear address of the committed Enclave pages are guard pages, from the
+Enclave code's point of view. Enclave code access to [0, `ENCLAVE_START`-1], or
+[`ENCLAVE_END`+1, 0 +`SECS.SIZE`-1], will trigger a Page Fault. Enclave code
+access to linear address of `SECS.SIZE` and beyond can success if the OS Page
+Table contains a valid entry covering the linear address. For the Enclave code
+to access shared memory between the Enclave and the host part of the
+application, for example, the ECALL/OCALL data marshalling buffers or host-side
+code provided shared memory, those shared memory must reside above `SECS.SIZE`,
+not in the address range lower than `ENCLAVE_START`.
+
+![0-base layout](0-base.png "0-base Enclave Memory Layout")
+
+In a typical SGX application, `ENCLAVE_START` is selected by the OS memory
+manager, and `SECS.BASEADDR` is set to match `ENCLAVE_START`. In SGX, the
+committed Enclave page's linear address offset from SECS.BASEADDR is included in
+the Enclave measurement. In the case that `SECS.BASEADDR` = `ENCLAVE_START`, the
+OS memory manager can place [`ENCLAVE_START`, `ENCLAVE_END`] at different linear
+address, and the Enclave measurement will not be affected. With `SECS.BASEADDR`
+set to 0, `ENCLAVE_START` must be a predetermined linear address, selected at
+Enclave signing time. Both Windows and Linux memory manager APIs support
+reserving memory at preferred linear address. Those APIs can be used to support
+placing the 0-base SGX Enclave in the memory location required. Typically, the
+OS memory manager does not allow reserving address below certain threshold. For
+Linux, the threshold is typically 4KB or 64KB, and configurable by the system
+admin. For Windows, the threshold is 64KB. Selecting a `ENCLAVE_START` value
+much higher than the typical threshold, especially for Linux, makes sure the
+enclave can be loaded on most systems.  The Enclave Loader does not need to
+reserve the linear address range between `SECS.BASEADDR`=0 and `SECS.SIZE` for
+the Enclave. It should reserve the memory range starting from `ENCLAVE_START`,
+in this case, predefined at Enclave signing time, with a size smaller than
+`SECS.SIZE`.
+
+It's possible to support multiple 0-base Enclaves in a single process. The SGX
```

Comment:
We should verify that Windows permits this and explicitly scope out supporting Windows if it does not.

Reply:
Added a paragraph in the updated PR to clarify that no POC was implemented on Windows yet and Windows support would have to wait, if the existing Windows kernel APIs for SGX Enclave won't allow "0-base".